honeybadgermpc/mpc.py

```diff
@@ -161,129 +188,254 @@ def _send(j, o):
+    class GFElementFuture(asyncio.Future):
+        """Represents a future for GFElement. Allows for arithmetic operations
+        to be stacked on top of the future value of this for when the value is resolved
 
-    def _binop_field(fut, other, op):
-        assert type(other) in [ShareFuture, GFElementFuture, Share, GFElement]
-        if isinstance(other, ShareFuture) or isinstance(other, Share):
-            res = ShareFuture()
-        elif isinstance(other, GFElement) or isinstance(other, GFElementFuture):
-            res = GFElementFuture()
+        TODO: Add more methods from GFElement to GFElementFuture
+            https://github.com/initc3/HoneyBadgerMPC/issues/245
+        """
 
-        if isinstance(other, asyncio.Future):
-            def cb(f): return res.set_result(op(fut.result(), other.result()))
-            asyncio.gather(fut, other).add_done_callback(cb)
-        else:
-            def cb(f): return res.set_result(op(fut.result(), other))
-            fut.add_done_callback(cb)
-        return res
+        def __add__(self, other):
+            return self.__binop_field(other, lambda a, b: a + b)
 
-    class GFElementFuture(asyncio.Future):
-        def __add__(self, other): return _binop_field(
-            self, other, lambda a, b: a + b)
+        __radd__ = __add__
 
-        def __sub__(self, other): return _binop_field(
-            self, other, lambda a, b: a - b)
+        def __sub__(self, other):
+            return self.__binop_field(other, lambda a, b: a - b)
 
-        def __mul__(self, other): return _binop_field(
-            self, other, lambda a, b: a * b)
+        def __rsub__(self, other):
+            return self.__binop_field(other, lambda a, b: -a + b)
+
+        def __mul__(self, other):
+            return self.__binop_field(other, lambda a, b: a * b)
+
+        def __binop_field(self, other, op):
+            """Stacks the application of a function to the resolved value of
+            this future GFElement
+            """
+
+            if isinstance(other, int):
+                other = context.field(other)
+
+            if not isinstance(other, (GFElement, GFElementFuture)):
+                return NotImplemented
+
+            res = GFElementFuture()
+
+            if isinstance(other, GFElementFuture):
+                asyncio.gather(self, other).add_done_callback(
+                    lambda _: res.set_result(op(self.result(), other.result())))
+            else:
+                self.add_done_callback(
+                    lambda _: res.set_result(op(self.result(), other)))
+
+            return res
 
     class Share(object):
+        """Represents a local share of a secret-shared GFElement
+        """
+
         def __init__(self, v, t=None):
             # v is the local value of the share
             if type(v) is int:
                 v = context.field(v)
-            assert type(v) is GFElement
+            assert isinstance(v, (GFElement, GFElementFuture))
```

Comment:
So, the main idea is that `ShareFuture` objects allow you to stack up `Share` arithmetic operations on each other on potentially unresolved values. I'm not entirely sure when you'd want to explicitly use a share with a `GFElementFuture`, but @smkuls  mentioned that it would be a good method to add in during the refactor.